docs: add postgres reference section to docs (close #4440) (#4471)
hasura/graphql-engine: docs/graphql/manual/guides/postgres/functions.rst

.. meta::
   :description: Use Postgres functions with Hasura
   :keywords: hasura, docs, postgres, functions

.. _postgres_functions:

Postgres functions
==================

.. contents:: Table of contents
  :backlinks: none
  :depth: 1
  :local:

Introduction
------------

`Postgres functions <https://www.postgresql.org/docs/current/sql-createfunction.html>`__ allow you to customize your database schema by defining a set of operations that can include several statements such as declarations, assignments and conditional workflows. 

> Postgres functions are similar to views but are used when we have arguments.

Reviewer: ```suggestion
Postgres functions are similar to views but allow more imperative computation and can take arguments.
```

.. note::

  For more information on Postgres functions, please refer to the `Postgres documentation <https://www.postgresql.org/docs/current/sql-createfunction.html>`__.

> Examples

Reviewer: It seems both examples are trigger functions, we should perhaps have an example of a trackable function with some arguments? (can copy straight from schema section in docs)

Reviewer: In fact, since we also have a section on `triggers`, we should probably not talk about trigger functions at all.

      BEGIN
      SELECT author.is_active INTO active_author FROM "authors" author WHERE author.id = NEW."author_id";
      IF active_author != TRUE THEN
          RAISE EXCEPTION 'Author must be active';
      END IF;
      RETURN NEW;
      END;
      $BODY$ LANGUAGE plpgsql;

Let's break this function apart:

- Function name: ``check_author_active``
- Parameters: This function doesn't have parameters (see ``()``)
- Return type: ``trigger``
- Variable declaration: A variable called ``active_author`` is declared
- Function body: Block between ``BEGIN`` and ``END`` checking if the author for whom the article is to be inserted is active
- Response: The response (``$BODY$``) is returned in the ``slpgsql`` language

.. _pg_function_example_two:

**Refresh a materialized view:**

The objective of the following function is to refresh a :ref:`materialized view <pg_materialized_view_example>`.

.. code-block:: plpgsql

  CREATE FUNCTION refresh_materialized_view()
    RETURNS trigger AS $BODY$
    BEGIN
    REFRESH MATERIALIZED VIEW popular_active_authors;
    RETURN NULL;
    END;
    $BODY$ LANGUAGE plpgsql;

Postgres functions & Hasura
---------------------------

> Postgres functions can be used to extend your Hasura GraphQL API, and they can also be exposed via the same.

Reviewer: ```suggestion
Postgres functions can be exposed in Hasura's GraphQL Schema. They are typically used for performing custom business logic in the database.
```
